boa/src/syntax/ast/node/operator/unary_op/mod.rs

```diff
@@ -76,13 +76,16 @@ impl Executable for UnaryOp {
             }
             op::UnaryOp::Not => self.target().run(context)?.not(context)?.into(),
             op::UnaryOp::Tilde => {
-                let num_v_a = self.target().run(context)?.to_number(context)?;
-                Value::from(if num_v_a.is_nan() {
-                    -1
+                let num_v_a = self.target().run(context)?.to_numeric_number(context)?;
+                if num_v_a.is_nan() || num_v_a.is_infinite() {
+                    Value::from(-1) // special case for inf or nan
+                } else if let Some(num_bigint) = self.target.run(context)?.as_bigint() {
+                    Value::from(JsBigInt::not(num_bigint))
+                // add bigint support
                 } else {
-                    // TODO: this is not spec compliant.
-                    !(num_v_a as i32)
-                })
+                    let masked = (num_v_a as i64) & 0x00000000ffffffff; // converts float to i32 following spec to ignore MSB using mask
+                    Value::from(!(masked as i32)) // Nots i32 conversion and creates value from this
+                }
```

Comment:
```
Let expr be the result of evaluating UnaryExpression.
2. Let oldValue be ? ToNumeric(? GetValue(expr)).
3. Let T be Type(oldValue).
4. Return ! T::bitwiseNOT(oldValue).
```

We should call `to_numeric` instead of `to_numeric_number` (which is a specialization of `to_numeric` for performace reasons). And then match on the numeric structure.

for bigint match arm we should call `JsBigInt::not()`.
And for number arm we should have a associated function [`Number::not(x: f64)`](https://tc39.es/ecma262/#sec-numeric-types-number-bitwiseNOT)
```
let x = f64_to_int32(x);
return !x;
```